src/ai/backend/manager/idle.py

```diff
@@ -307,24 +305,148 @@ class UtilizationIdleChecker(BaseIdleChecker):
     """
 
     name: ClassVar[str] = "utilization"
-    default_cpu_util_threshold: ClassVar[float] = 30.0
-    default_accelerator_util_threshold: ClassVar[float] = 10.0
-    default_initial_grace_period: ClassVar[float] = 300.0  # allow first 5 minutes to be idle
+    cpu_util_series: List[float] = []
+    mem_util_series: List[float] = []
+    cuda_mem_util_series: List[float] = []
 
-    async def populate_config(self, config: Mapping[str, Any]) -> None:
-        pass
+    _evhandlers: List[EventHandler[None, AbstractEvent]]
 
-    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
-        # last_stat = session['last_stat']
-        # TODO: implement
-        return True
+    async def populate_config(self, raw_config: Mapping[str, Any]) -> None:
+        self.cpu_threshold = int(raw_config["resource-thresholds"]["cpu"]["average"])
+        self.mem_threshold = int(raw_config["resource-thresholds"]["mem"]["average"])
+        self.cuda_mem_threshold = int(raw_config["resource-thresholds"]["cuda_mem"]["average"])
+        self.threshold_condition = str(raw_config["thresholds_check_condition"]["condition"])
+        self.window = raw_config["resource-thresholds"]["cpu"]["window"]
+
+        log.info(
+            f"ExecutionIdleChecker: default cpu idle_timeout = {self.cpu_threshold} pct; \
+              Mem idle execution = {self.mem_threshold} %, \
+              and observed moving window = {self.window}"
+        )
 
     async def update_app_streaming_status(
         self,
         session_id: SessionId,
         status: AppStreamingStatus,
     ) -> None:
-        pass
+        if status == AppStreamingStatus.HAS_ACTIVE_CONNECTIONS:
+            await self._disable_timeout(session_id)
+        elif status == AppStreamingStatus.NO_ACTIVE_CONNECTIONS:
+            await self._update_timeout(session_id)
+
+    async def _disable_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._disable_timeout({session_id})")
+        await self._redis.set(
+            f"session.{session_id}.last_execution", "0", exist=self._redis.SET_IF_EXIST
+        )
+
+    async def _update_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._update_timeout({session_id})")
+        t = await self._redis.time()
+        await self._redis.set(
+            f"session.{session_id}.last_execution",
+            f"{t:.06f}",
+            expire=float(sum(self.cpu_util_series) / len(self.cpu_util_series))
```

Comment:
I didn't understand this line. Why the expiration of redis key is dependent on CPU utilization history?

Reply:
I had just a draft idea of using CPU utilization as a proxy for time measure, to make the expiration record more dynamic according to resource usage. The more CPU utilized the more time is given for expiration record. However, I didn't come out with a solution of scale range definition that maps cpu usage to time. Now, I consider to set expiration to 86400 seconds as the default value for storing record.

Reply:
Aha. I think your idea is great. The only problem here is that there should be a situation that CPU utilization is turned off. Some admins may want to garbage collect session based only on the GPU utilization. In that case, the CPU utilization has no meaning in terms of dynamic expiration time.

If you can introduce some logic of generalized utilization or something else, I agree with your idea of dynamic expiration time. Or, you can make it configurable by introducing a Etcd key.